tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -432,49 +434,129 @@ impl StackGraphLanguage {
     /// nodes and edges in `stack_graph`.  Any new nodes that we create will belong to `file`.
     /// (The source file must be implemented in this language, otherwise you'll probably get a
     /// parse error.)
-    pub fn build_stack_graph_into(
-        &self,
-        stack_graph: &mut StackGraph,
+    pub fn build_stack_graph_into<'a>(
+        &'a self,
+        stack_graph: &'a mut StackGraph,
+        file: Handle<File>,
+        source: &'a str,
+        globals: Variables<'a>,
+        cancellation_flag: &'a dyn CancellationFlag,
+    ) -> Result<(), LoadError> {
+        self.builder_into_stack_graph(stack_graph, file, source)
+            .build(globals, cancellation_flag)
+    }
+
+    /// Create a builder that will the graph construction rules for this language against a source
+    /// file, creating new nodes and edges in `stack_graph`.  Any new nodes that we create will
+    /// belong to `file`.  (The source file must be implemented in this language, otherwise you'll
+    /// probably get a parse error.)
+    pub fn builder_into_stack_graph<'a>(
+        &'a self,
+        stack_graph: &'a mut StackGraph,
+        file: Handle<File>,
+        source: &'a str,
+    ) -> Builder<'a> {
+        Builder::new(self, stack_graph, file, source)
+    }
+}
+
+pub struct Builder<'a> {
+    sgl: &'a StackGraphLanguage,
+    stack_graph: &'a mut StackGraph,
+    file: Handle<File>,
+    source: &'a str,
+    graph: Graph<'a>,
+    remapped_local_ids: HashMap<usize, NodeID>,
+    injected_node_count: usize,
```

Comment:
We count the injected nodes, which will be the lowest node indices (all injections happen before the TSG is executed) and are treated special when processing the graph.